velox/functions/prestosql/aggregates/ApproxPercentileAggregate.cpp

```diff
@@ -473,43 +473,56 @@ class ApproxPercentileAggregate : public exec::Aggregate {
       const SelectivityVector& rows,
       const BaseVector& vec) {
     DecodedVector decoded(vec, rows);
-    VELOX_USER_CHECK(
-        decoded.isConstantMapping(),
-        "Percentile argument must be constant for all input rows");
     bool isArray;
     const double* data;
     vector_size_t len;
+    std::vector<bool> isNull;
     auto i = decoded.index(0);
     if (decoded.base()->typeKind() == TypeKind::DOUBLE) {
       isArray = false;
-      data =
-          decoded.base()->asUnchecked<ConstantVector<double>>()->rawValues() +
-          i;
+      auto baseVector = decoded.base();
+      if (baseVector->encoding() == VectorEncoding::Simple::FLAT) {
+        data = baseVector->asUnchecked<FlatVector<double>>()->rawValues() + i;
+      } else {
+        // ConstantVector
+        data = baseVector->asUnchecked<ConstantVector<double>>()->rawValues();
+      }
       len = 1;
+      isNull = {baseVector->isNullAt(i)};
     } else if (decoded.base()->typeKind() == TypeKind::ARRAY) {
       isArray = true;
       auto arrays = decoded.base()->asUnchecked<ArrayVector>();
       VELOX_USER_CHECK(
           arrays->elements()->isFlatEncoding(),
           "Only flat encoding is allowed for percentile array elements");
       auto elements = arrays->elements()->asFlatVector<double>();
-      data = elements->rawValues() + arrays->offsetAt(i);
+      auto startIndex = arrays->offsetAt(i);
```

Comment:
Can you extract the common code to convert percentiles into `(const double* data, vector_size_t len, std::vector<bool> isNull)`?

Reply:
extracted into `extractPrecentiles`